lib/gauge_random.cu

```diff
@@ -11,196 +11,183 @@
+      switch (U.Reconstruct()) {
+      case QUDA_RECONSTRUCT_NO: genGauss<Float, QUDA_RECONSTRUCT_NO, true>(U, rngstate, beta); break;
+      case QUDA_RECONSTRUCT_12: genGauss<Float, QUDA_RECONSTRUCT_12, true>(U, rngstate, beta); break;
+      case QUDA_RECONSTRUCT_8: genGauss<Float, QUDA_RECONSTRUCT_8, true>(U, rngstate, beta); break;
+      default: errorQuda("Reconstruction type %d of gauge field not supported", U.Reconstruct());
       }
-      void postTune(){
-	  //gf.restore();
-	  arg.rngstate.restore();
+    } else if (U.LinkType() == QUDA_MOMENTUM_LINKS) { // generate Gaussian distributed su(3) field
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Creating Gaussian distrbuted momentum field\n");
+      switch (U.Reconstruct()) {
+      case QUDA_RECONSTRUCT_NO: genGauss<Float, QUDA_RECONSTRUCT_NO, false>(U, rngstate, beta); break;
+      case QUDA_RECONSTRUCT_10: genGauss<Float, QUDA_RECONSTRUCT_10, false>(U, rngstate, beta); break;
+      default: errorQuda("Reconstruction type %d of gauge field not supported", U.Reconstruct());
       }
-
-    }; 
-
-  template<typename Float, typename Gauge>
-  void genGauss(const Gauge dataDs, GaugeField& data, RNG &rngstate) {
-      GaugeGaussArg<Gauge> arg(dataDs, data, rngstate);
-      GaugeGauss<Float,Gauge> gaugeGauss(arg, data);
-      gaugeGauss.apply(0);
-
     }
-
-
-
-  template<typename Float>
-  void gaugeGauss(GaugeField &dataDs, RNG &rngstate) {
-
-      if(dataDs.Reconstruct() == QUDA_RECONSTRUCT_NO) {
-	  typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_NO>::type Gauge;
-	  genGauss<Float>(Gauge(dataDs), dataDs, rngstate);
-      }else if(dataDs.Reconstruct() == QUDA_RECONSTRUCT_12){
-	  typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_12>::type Gauge;
-	  genGauss<Float>(Gauge(dataDs), dataDs, rngstate);
-      }else if(dataDs.Reconstruct() == QUDA_RECONSTRUCT_8){
-	  typedef typename gauge_mapper<Float,QUDA_RECONSTRUCT_8>::type Gauge;
-	  genGauss<Float>(Gauge(dataDs), dataDs, rngstate);
-      }else{
-	  errorQuda("Reconstruction type %d of origin gauge field not supported", dataDs.Reconstruct());
-      }
-
   }
 
-#endif
-
-  void gaugeGauss(GaugeField &dataDs, RNG &rngstate) {
-
-#ifdef GPU_GAUGE_TOOLS
-
-      if(dataDs.Precision() == QUDA_HALF_PRECISION){
-	  errorQuda("Half precision not supported\n");
-      }
+  void gaugeGauss(GaugeField &U, RNG &rngstate, double beta)
+  {
+    if (!U.isNative()) errorQuda("Order %d with %d reconstruct not supported", U.Order(), U.Reconstruct());
+    if (U.Ncolor() != 3) errorQuda("Nc = %d not supported", U.Ncolor());
 
-      if (!dataDs.isNative())
-	  errorQuda("Order %d with %d reconstruct not supported", dataDs.Order(), dataDs.Reconstruct());
+    switch (U.Precision()) {
+    case QUDA_DOUBLE_PRECISION: gaugeGauss<double>(U, rngstate, beta); break;
+    case QUDA_SINGLE_PRECISION: gaugeGauss<float>(U, rngstate, beta); break;
```

Comment:
We should add a version for `QUDA_HALF_PRECISION` and `QUDA_QUARTER_PRECISION` which creates a single precision field then copies it into a half/quarter field, for completeness.

Reply:
I'm not sure where this will ever be needed.  It's needed for the quark fields since this is used in the MG setup.  Can't see why one would ever need to create a half or quarter precision gauge field.

Reply:
Ah, fair, agreed!